### ISSUE TITLE
Stable torch.sort and torch.argsort

### ISSUE THREAD

**Issue by agadetsky** [pytorch/pytorch#38681]
## 🚀 Feature
Add stable version of `torch.sort` and `torch.argsort`. Stable sort algorithms sort repeated elements in the same order that they appear in the input.

## Motivation

In some applications we need to preserve the order of equal elements while sorting.
## Alternatives

Now we can do:

```
def stable_argsort(arr, dim=-1, descending=False):
    arr_np = arr.detach().cpu().numpy()
    if descending:
        indices = np.argsort(-arr_np, axis=dim, kind='stable')
    else:
        indices = np.argsort(arr_np, axis=dim, kind='stable')
    return torch.from_numpy(indices).long().to(arr.device)
```

and

```
def stable_sort(arr, dim=-1, descending=False):
    arr_np = arr.detach().cpu().numpy()
    if descending:
        indices = np.sort(-arr_np, axis=dim, kind='stable')
    else:
        indices = np.sort(arr_np, axis=dim, kind='stable')
    return torch.from_numpy(indices).as_type(arr)
```

But I think that it will be cool to have this functionality in `torch.sort` and `torch.argsort`.

**Comment by guol-fnst** [comment]
Hi @agolynski 
What about adding a new parameter <stable='true|false' > ?
Just like `torch.argsort(input, dim=-1, descending=False, stable=True)`

**Comment by agadetsky** [comment]
@nikitaved seems like #39744 is almost finished. Will you be able to work on stable sort PR?

**Comment by nikitaved** [comment]
Sure, once the PR is approved!

**Comment by agadetsky** [comment]
@nikitaved seems like #39744 is finished. Will you be able to work on adding stable sort/argsort?

**Comment by nikitaved** [comment]
I will, but not until the beginning of next week. The CPU part is straightforward, the CUDA part will certainly require port, unless `stable_sort` is a separate routine. But, I presume, it is better to be just `sort` with an additional flag.

**Comment by agadetsky** [comment]
@nikitaved thank you, will be waiting for your PR.

**Comment by nikitaved** [comment]
@agadetsky , the CPU stable sort is available on CPU already. The CUDA part will need some research in order to find out whether there are efficient extension of bitonic sort to stability, or maybe there are other efficient algorithms.

**Comment by agadetsky** [comment]
@nikitaved thank you very much! Will be waiting for cuda version as well.

**Comment by agadetsky** [comment]
@nikitaved are you going to implement stable sort for GPU?

**Comment by agadetsky** [comment]
@ngimel why you closed this issues? It seems like stable sort and stable argsort are not implemented for GPU yet.

**Comment by ngimel** [comment]
Stable sort is implemented for gpu.

**Comment by agadetsky** [comment]
@ngimel thank you for your quick response. But docs for 1.9 say that stable sort works only for CPU for now. See https://pytorch.org/docs/stable/generated/torch.sort.html?highlight=sort#torch.sort

**Comment by ngimel** [comment]
That's the error in docs, thanks for pointing it out! We'll accept a PR fixing it.
